Add the option to enable default Pod Security Configuration

Foxlik opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md and developer guide https://git.k8s.io/community/contributors/devel/development.md
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?** 
/kind feature

**What this PR does / why we need it**:
By default only Pods in annotated namespaces are checked for following Pod Security Standards. This change adds the option to reverse the behavior, where Pods in all namespaces not exmpted are checked.

**Special notes for your reviewer**:
This is my first PR to kubespray and this looks like a quick and easy win to me, but I might be mistaken. Thanks for any feedback.

I'm not sure if any other namespaces should be on the default exempt list.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Add the option to enable default Pod Security Configuration
```

cristicalin commented:
Hi @Foxlik , thanks for your contribution, please ensure to sign the CLA so it can be accepted.

oomichi commented:
@alegrey91 Could you take a look at this pull request? I guess you might be interested.

alegrey91 commented:
@oomichi it looks really interesting. Thank for letting me know.
At first glance, seems that the `PodSecurityConfiguration` manifest is never applied to the cluster.
For the rest, it seems to work.
@Foxlik please fix the PR problems asap.

alegrey91 commented on roles/kubernetes/control-plane/vars/main.yaml:
```diff
@@ -1,3 +1,3 @@
 ---
 # list of admission plugins that needs to be configured
-kube_apiserver_admission_plugins_needs_configuration: [EventRateLimit]
+kube_apiserver_admission_plugins_needs_configuration: [EventRateLimit, PodSecurity]
```
This variable has been created to ensure that admission plugins that need to be configured, could be provided with this.
In this case, I think it is not necessary to add `PodSecurity` here because the plugin doesn't need configuration to work.
Here are some examples on how to use it:
https://github.com/kubernetes-sigs/kubespray/search?q=kube_apiserver_admission_plugins_needs_configuration

Foxlik replied:
@alegrey91  Getting this [PodSecurityConfiguration](https://github.com/kubernetes-sigs/kubespray/pull/9017/files#diff-b0903679f8b21f320860f39e46a4c69163d8eaa91eb8a9f8ec8d280fc3872de1) file created on the node is the sole reason I got into creating this PR. It's an admission plugin so I thought this was a good place to get it done. If there is a better place or if you think it's more clear to do it in other way, I'm ready to fix that. But the plugin needs the config to work.
I think this also answers your [other comment](https://github.com/kubernetes-sigs/kubespray/pull/9017#issuecomment-1182550241) about the PodSecurityConfiguration not being applied to the cluster.

The patch definitely does what's advertised - enable default PodSecurity for whole cluster and when configured, prevent insecure Pods from being spawned. Trust me, it forced me to fix quite a few apps ;).

Foxlik replied:
@cristicalin i've added a condition around the PodSecurity in here. I did it the most naive way. Not sure if it's up to standards. Please check. 
The original idea was to get the file created always, just empty when `kube_pod_security_use_default` is false.

alegrey91 replied:
@Foxlik I got your point. 

So, you can leave the code in this way:
```
kube_apiserver_admission_plugins_needs_configuration: [EventRateLimit,PodSecurity]
```
You should probably rename the file [`roles/kubernetes/control-plane/templates/podsecurity.v1beta2.yaml.j2`](https://github.com/kubernetes-sigs/kubespray/pull/9017/files#diff-b0903679f8b21f320860f39e46a4c69163d8eaa91eb8a9f8ec8d280fc3872de1) to `roles/kubernetes/control-plane/templates/podsecurity.yaml.j2`.

This because the [`admission-controls.yaml.j2`](https://github.com/kubernetes-sigs/kubespray/blob/master/roles/kubernetes/control-plane/templates/admission-controls.yaml.j2) will grab the name from the variable `kube_apiserver_admission_plugins_needs_configuration` and use it as name of the `AdmissionConfiguration` plugin list.

E.g.
```

apiVersion: apiserver.config.k8s.io/v1
kind: AdmissionConfiguration
plugins:
- name: EventRateLimit
  path: /path/to/eventratelimit.yaml
- name: PodSecurity                        # extracted from plugin variable
  path: /path/to/podsecurity.yaml          # plugin | lower
```

Ping me if you have some doubt about the explanation.
Anyway, you did a great work!

Foxlik commented on roles/kubernetes/control-plane/defaults/main/main.yml:
```diff
@@ -104,6 +104,19 @@ kube_apiserver_admission_control_config_file: false
+kube_pod_security_exemptions_namespaces:
+  - kube-system
+
+
```
Removed the empty line.

Foxlik commented:
I finally got the CLA sorted out. Reverted the latest changes, but added an explanation for the empty file. If this looks good to you, please set it to squash before merge.
Thanks.

floryut commented:
> I finally got the CLA sorted out. Reverted the latest changes, but added an explanation for the empty file. If this looks good to you, please set it to squash before merge. Thanks.

Thank you, could you please rebase master though ? Otherwise CI won't  be happy :)

Foxlik commented:
> > I finally got the CLA sorted out. Reverted the latest changes, but added an explanation for the empty file. If this looks good to you, please set it to squash before merge. Thanks.
> 
> Thank you, could you please rebase master though ? Otherwise CI won't be happy :)

guess I should've said I've done so. anyway, rebased again.

cristicalin commented:
Thanks for this work @Foxlik !

/approve
/ok-to-test

cristicalin commented:
/cc @floryut @oomichi

floryut reviewed (approved):
@Foxlik Sorry I totaly forgot this one, nice work on this
/lgtm
